test/api/bucket/api_test.go

```diff
@@ -0,0 +1,205 @@
+package bucket
+
+import (
+	"bytes"
+	"compress/gzip"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/konveyor/tackle2-hub/api"
+	"github.com/konveyor/tackle2-hub/binding"
+	"github.com/konveyor/tackle2-hub/test/assert"
+)
+
+func TestBucketCRUD(t *testing.T) {
+	for _, sample := range Samples {
+		t.Run("Bucket CRUD Test", func(t *testing.T) {
+
+			expectedBucket := api.Bucket{
+				Path: sample.Path,
+			}
+
+			// Create a new bucket.
+			assert.Must(t, Bucket.Create(&expectedBucket))
+
+			// Get all the buckets.
+			gotBuckets, err := Bucket.List()
+			if err != nil {
+				t.Errorf(err.Error())
+			}
+
+			// Find for the specific bucket and compare Paths as it is a unique value.
+			for _, gotBucket := range gotBuckets {
+				if gotBucket.ID == expectedBucket.ID {
+					if gotBucket.Path != expectedBucket.Path {
+						t.Errorf("Difference in Path between the buckets %v and %v", gotBucket.Path, expectedBucket.Path)
+					}
+				}
+			}
+
+			// Inject Expected Buckets's ID into the BucketRoot.
+			pathForBucket := binding.Path(api.BucketRoot).Inject(binding.Params{api.ID: expectedBucket.ID})
+
+			// Get specific bucket.
+			gotBucket := api.Bucket{}
+			err = Client.Get(pathForBucket, &gotBucket)
+			if err != nil {
+				t.Errorf(err.Error())
+			}
+
+			// Compare bucket Paths.
+			if gotBucket.Path != expectedBucket.Path {
+				t.Errorf("Difference in Path between the buckets %v and %v", gotBucket.Path, expectedBucket.Path)
+			}
+
+			// Delete bucket.
+			assert.Must(t, Bucket.Delete(pathForBucket))
+		})
+	}
+}
+
+func TestBucketFile(t *testing.T) {
```

Comment:
It may turn out to be a rehash of a previous conversation: if we are going to loop through all of the "buckets" and create + delete them, why wouldn't we do all of the operations on created buckets before deleting them. For example, instead of having `TestBucketCRUD` and `TestBucketFile` that both loop through all Samples and have slight differences from there:

1. Create a bunch of buckets, verify each was created correctly
2. List the buckets and make sure that they were all created correctly
3. Do File/Directory related operations on our created buckets
4. Delete the buckets, verify each was deleted correctly
5. Done

I feel like I've had this discussion before and I'm hoping I am not leading you astray. Feel free to point me to any related commentary if it exists.

Reply:
Thank you for the review and I think we can merge them into one function and that it will reduce the number of LOC and the variables we might be using but this might increase the complexity of the code and also the number of dump variables in active state at that time. Also might decrease readability to the future users. What do you think?

Reply:
We may be able to leverage something like https://pkg.go.dev/testing#hdr-Subtests_and_Sub_benchmarks to make it so we can clearly separate these tests (for readability) and allow them to run sequentially to prevent us from thrashing the API with create/delete bucket requests. I'll look through what you have and see if that would help us.